### DISCUSSION TITLE
How to generate rays_<datasetname>.dat to train

### DISCUSSION THREAD

**Discussion by LuckyOne09** [NVlabs/instant-ngp#797] [Q&A] [answered]
Hello, thank you for your great and complete work first. But I come across a problem when try to reconstruct a scene by ngp.
I am reconstructing a scene captured by cameras with different intrinsic parameters. But I notice ngp assumes using same camera in transform.json.
Following [this hint](https://github.com/NVlabs/instant-ngp/issues/333) , I want to pre-compute all the rays,  but I dont know how to generate that, could you give me some guidelines?

**Comment by Tom94** [comment, marked as answer]
Hi there, you can actually provide per-frame intrinsics by now -- so you don't need to precompute rays. Simply bundle the same parameters that are global by default ("cx", "cy", etc.) within the frame's scope (alongside "file_path" and so on). All frames with such custom intrinsics will use those instead of the global ones. Cheers!

**Reply by LuckyOne09** [reply]
Thank you for your quick reply. 
You mean I can give a transform.json like this? 
```
{
  "aabb_scale": 1,
  "frames": [
    {
      "camera_angle_x": 0.8681504423609672,
      "camera_angle_y": 1.107881864900016,
      "fl_x": 1553.1880767941186,
      "fl_y": 1552.040159329971,
      "k1": 0,
      "k2": 0,
      "p1": 0,
      "p2": 0,
      "cx": 723.5649,
      "cy": 965.10596,
      "w": 1440.0,
      "h": 1920.0,
      "file_path": "ngp_images/VID_00/127.jpg",
      "sharpness": 30.71202072602876,
      "transform_matrix": [
        [
          0.9888559479731736,
          -0.03915492456075905,
          -0.14363427878023233,
          -0.3039944156594588
        ],
        [
          0.044095460883605074,
          0.9985345284198722,
          0.031374924426454426,
          -0.4993807267181046
        ],
        [
          0.14219530402773042,
          -0.037358900357806826,
          0.989133362128948,
          -0.3000137377426212
        ],
        [
          0.0,
          0.0,
          0.0,
          1.0
        ]
      ]
    },
    {
      "camera_angle_x": 0.868,
      "camera_angle_y": 1.107,
      "fl_x": 1553.18,
      "fl_y": 1552.04,
      "k1": 0,
      "k2": 0,
      "p1": 0,
      "p2": 0,
      "cx": 723.5,
      "cy": 965.1,
      "w": 1440.0,
      "h": 1920.0,
      "file_path": "ngp_images/VID_00/126.jpg",
      "sharpness": 30.71202072602876,
      "transform_matrix": [
        [
          0.9879648339176126,
          -0.05551373288009035,
          -0.14437351697545522,
          -0.3196259070879651
        ],
        [
          0.04986237349924451,
          0.9978525087901724,
          -0.04247486798308795,
          -0.41966387529599847
        ],
        [
          0.14642141459214883,
          0.03476486966575435,
          0.9886112346043601,
          -0.30368980273279744
        ],
        [
          0.0,
          0.0,
          0.0,
          1.0
        ]
      ]
    }
  ],
  "aabb": [
    [
      -4.278669834136963,
      -1.87978994846344,
      -4.775090217590332
    ],
    [
      3.4867000579833984,
      2.926409959793091,
      5.331930160522461
    ]
  ]
}
```

**Reply by Tom94** [reply]
Yes

**Reply by fishfishson** [reply]
Hi @Tom94  I find in **run.py** file, ngp uses global fov to render images. Can I use per frame intrinsic parameters like parsing the above json file to render images? If so, could you pls tell me how to implement it?

**Reply by Tom94** [reply]
You can use the Python bindings (by modifying `run.py`) to change testbed's FoV setting depending on which image you're rendering out. See `python_api.cu` for all the available bindings (you should be able to find the ones controlling fov there).

**Reply by fishfishson** [reply]
like testbed.set_fov_xy = [angle_x, angle_y]?

**Reply by Mabel0403** [reply]
> like testbed.set_fov_xy = [angle_x, angle_y]?

Hi @fishfishson, I have similar problems as you. I used images with different intrinsics, getting a NeRF model and base.ingp. Then, I tried to render an image by giving a .json. However, the image which run.py outputed isn't from the perspective I need. Can I discuss this problem with you in detail? Thanks a lot!
